### PR TITLE
Add missing RelationThrough#beforeInsert' and Join typings

### PR DESCRIPTION
**Pull request by rgtk**
http://vincit.github.io/objection.js/#relationmappings

### CODE REVIEW

File: typings/objection/index.d.ts [hunk 1]
```diff
@@ -198,6 +213,7 @@ declare namespace Objection {
     to: string | Reference | (string | Reference)[];
     modelClass?: ModelClass<any> | string;
     extra?: string[] | object;
+    beforeInsert?: (model: Model, context: QueryContext) => void;
   }
```
Comment (koskimas) [inline typings/objection/index.d.ts]: This method can be async, so maybe the return value should be `void | Promise<any>` or something? I'm not sure how other async hooks have been typed.

**Comment by koskimas** [comment]
It seems that you have duplicates of your old commits in this PR. Please rebase master to remove those.

**Comment by rgtk** [comment]
Updated and added another missing typing.

File: typings/objection/index.d.ts [hunk 2]
```diff
@@ -213,6 +213,7 @@ declare namespace Objection {
     to: string | Reference | (string | Reference)[];
     modelClass?: ModelClass<any> | string;
     extra?: string[] | object;
+    beforeInsert?: (model: Model, context: QueryContext) => Promise<any> | void;
```
Comment (mattleff) [inline typings/objection/index.d.ts]: This should be `Promise<void> | void;`

**Comment by rgtk** [comment]
Feedback fix applied.

**Pull request merged**